### PR TITLE
Cmake config userspace

### PR DESCRIPTION
**Pull request by SebastianBoe**

### CODE REVIEW

**Comment by SebastianBoe** [comment]
CMake does CONFIG_USERSPACE slightly differently from KBuild. The difference is ...

- We do not have the build system create archive files for kernel space and user space.
- We do not have the linker script list what files are kernel space archives.

Instead we

- Create a build system variable that lists the kernel space archives
- Have the build system send information about what files are the kernel space archive files to the linker script through preprocessor defines.

The rationale is ...

We create fewer archive files and save some calls to ar. As a general rule we want the build system to do as little as possible.

Also, now the linker script no longer knows what the kernel space archive files are called, this kind of information belongs to the build system, the linker script being agnostic to this is a plus.

Unfortunately CONFIG_APPLICATION_MEMORY does not have any tests AFAICT (obj_validation does enables CONFIG_USERSPACE, but not CONFIG_APPLICATION_MEMORY), so I'm not sure if this part actually works (with CMake or KBuild).

**Comment by andrewboie** [comment]
I'm going to need to spend some time learning about CMake to properly review this patch.

It's true that CONFIG_APPLICATION_MEMORY doesn't have test cases yet, wanted to get system calls implemented first. In retrospect I probably could have implemented a case to at least verify that memory is in the proper location.

I can tell you what the end result needs to be:
- If CONFIG_APPLICATION_MEMORY is disabled, all toplevel globals go in kernel RAM area
- If CONFIG_APPLICATION_MEMORY is enabled, any toplevel globals declared in the application source, or under kernel's lib/ directory, or any third-party libraries linked in, go in the application memory space.

The idea behind using archives was to tell the linker that stuff had to end up in one place or another, using EXCLUDE_FILE and so forth: https://sourceware.org/binutils/docs/ld/Input-Section-Basics.html

What the final logic ended up being was that anything in libzephyr.a and kernel/lib.a were put in kernel memory, and everything else in app memory (assuming app memory is enabled). The libapplication.a archive ended up not being needed and it's continued presence was an oversight.

What I can and should do for you is write a test case that shows that declared variables are being routed to the correct sections. I will try to get to this early this week.

**Comment by SebastianBoe** [comment]
A test case would be very appreciated.

**Comment by SebastianBoe** [comment]
> The reason why kernel/lib.a is separate is that Kbuild uses some funny linker tricks (that I don't fully understand) that cause globals inside the library to be discarded if unused.

BTW: Although I haven't checked to make sure, I assume the difference between kernel and the other libs is that the kernel is --no-whole-archive'd, while the other libs are --whole-archive'd.

--no-whole-archive is the default way to use a library.

In a command line like this:

gcc main.c -lmy_archive

The linker will search my_archive for any symbols that it hasn't found yet, and then pull in the .o files in my_archive that resolves unresolved symbols. .o files in my_archive that didn't resolve any missing symbols will be ignored/discarded.

IMHO carefully organizing your codebase into object files and archives to take advantage of the above behaviour is no longer necessary now that LTO is relatively mature. So I suspect we could just enable LTO by default and not distinguish between --whole-archive and --no-whole-archive'd libs, but this will have to be investigated another day ...

**Comment by andrewboie** [comment]
Have a look at #4075 should do the trick.

File: tests/kernel/obj_validation/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+if(DEFINED ENV{BOARD})
```
Comment (nashif) [inline tests/kernel/obj_validation/CMakeLists.txt]: this test moved to tests/kernel/mem_protect/obj_validation

**Review by nashif — CHANGES_REQUESTED** [review]

File: include/misc/util.h [hunk 2]
```diff
@@ -172,6 +172,115 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
...
+ * Generates a sequence of code.
+ * Useful for generating code like;
+ *
+ * NRF_PWM0, NRF_PWM1, NRF_PWM2,
+ *
+ * @arg LEN: The length of the sequence. MUST be defined.
+ *
+ * @arg F(i, F_ARG): A macro function that accepts two arguments.
+ *
+ * @detailed_description: F is called repeatedly, the first argument
```
Comment (nashif) [inline include/misc/util.h]: not supported by doxygen?

File: include/misc/util.h [hunk 3]
```diff
@@ -172,6 +172,115 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
...
+ * is the index in the sequence, and the second argument is the third
+ * argument given to UTIL_LISTIFY.
+ *
+ * @example:
+ * #define FOO(i, _) NRF_PWM ## i ,
+ * { UTIL_LISTIFY(PWM_COUNT, FOO) }
+ * // The above two lines will generate the below:
+ * { NRF_PWM0 , NRF_PWM1 , }
+ *
+ * @NB: Calling UTIL_LISTIFY with undefined arguments has undefined
```
Comment (nashif) [inline include/misc/util.h]: ditto

**Review by nashif — CHANGES_REQUESTED** [review]
we need to make those doxygen comments compatible with Zephyr, otherwise this will cause build failures, this is not something we can override in one PR, it needs to be fixed.

File: include/misc/util.h (hunk 2 above)
Comment (dbkinder) [reply include/misc/util.h]: there is no @detailed_description directive.  Simply write this as:

      * @arg F(i, F_ARG): A macro function that accepts two arguments.
      *  F is called repeatedly. The first argument to F
      *  is the index in the sequence, and the second argument to F is the third
      *  argument given to UTIL_LISTIFY.

File: include/misc/util.h [hunk 4]
```diff
@@ -172,6 +172,115 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
...
+ *
+ * @arg LEN: The length of the sequence. MUST be defined.
+ *
+ * @arg F(i, F_ARG): A macro function that accepts two arguments.
+ *
+ * @detailed_description: F is called repeatedly, the first argument
+ * is the index in the sequence, and the second argument is the third
+ * argument given to UTIL_LISTIFY.
+ *
+ * @example:
```
Comment (dbkinder) [inline include/misc/util.h]: The @example directive expects a filename as a parameter (that's why you're getting the error "explicit link request to 'define' could not be resolved").  I'd suggest writing this as (note the 4 extra spaces after the * before the code example to tell doxygen to format it as a code block:

     * Example:
     *     #define FOO(i, _) NRF_PWM ## i ,
     *     { UTIL_LISTIFY(PWM_COUNT, FOO) }
     *     // The above two lines will generate the below:
     *     { NRF_PWM0 , NRF_PWM1 , } 
     *

File: include/misc/util.h (hunk 3 above)
Comment (dbkinder) [reply include/misc/util.h]: there is no ``@nb`` doxygen directive.  Perhaps you meant to use ``@note``

**Review by dbkinder — CHANGES_REQUESTED** [review]
Fixes for some doxygen errors.
P.S.  See https://www.stack.nl/~dimitri/doxygen/manual/commands.html

**Pull request merged**